CRM-21849: Inline Relationship Type Edit

mickadoo opened the pull request:
Overview
----------------------------------------

To make it easier to edit relationship types this PR adds an icon to show a popup for inline editing of relationship types.

Before
----------------------------------------

It was not possible to edit relationship types inline. You had to go to the `civicrm/admin/reltype` screen

![image](https://user-images.githubusercontent.com/6374064/37725749-93b38d7a-2d2b-11e8-98f8-f125b21a1ca9.png)

Calls to `getoptions` for Relationship with field 'relationship_type_id` would result in:

```
{

    "is_error": 1,
    "error_message": "The field 'relationship_type_id' has no associated option list."
}
```

After
----------------------------------------

The relationship type dropdown has an icon to allow inline editing.

![peek 2018-03-21 17-13](https://user-images.githubusercontent.com/6374064/37725652-599f37f6-2d2b-11e8-95bd-164b1dde6a69.gif)

The Relationship `getoptions` endpoint supports `relationship_type_id`

![image](https://user-images.githubusercontent.com/6374064/37726504-72f4a7ac-2d2d-11e8-8e7d-5322d3574ce4.png)

Technical Details
----------------------------------------

- Spec functions for `getoptions` were ignored if the concrete spec function was not implemented. I wanted to modify the spec, but I didn't want to override the generic `getoptions` function. I modified the generic function to support custom spec functions.
- Relationship type comparison in the [Relationship BAO](https://github.com/civicrm/civicrm-core/compare/master...compucorp:CRM-21849-inline-relationship-type-edit#diff-7291ff5a1c2bb9adbc0da0929ae9f1d8R542) used strict comparison, even though one type was an int (according to doc-block) and the other (the BAO property) was a string, so I had to cast one to an int
- When refreshing the options for relationship types we can't show all types. The original list is added [here](https://github.com/civicrm/civicrm-core/blob/8c9251b398a29d9b82996ae21825c9d6478fa576/CRM/Contact/Form/Relationship.php#L300) and uses some parameters to filter it. In order to get the same list, I needed to pass these to the `getoptions` call done from [crm.optionEdit.js](https://github.com/civicrm/civicrm-core/blob/3ef93345f277ee729cec391e02c619d161f5caa7/js/crm.optionEdit.js#L31) so I introduced `option_context` to the elements. If it exists it will be set as a property of the element (along with the other metadata, such as `data-api-entity` and `data-api-field`, and it will be passed along to the `getoptions` call.
- The `relationshipData` was previously populated in the backend using `CRM_Contact_BAO_Relationship::getContactRelationshipType` but this means that it stays static. To allow dynamically recreating `relationshipData` (after editing etc.) the generation was switched to the frontend

---

 * [CRM-21849: Add Link To Edit Relationship Types](https://issues.civicrm.org/jira/browse/CRM-21849)

seanmadsen commented:
(*CiviCRM Review Template WORD-1.1*)

<!-- In each category, change the word "Undecided" to "Pass" or "Issue". Add explanatory comments if prompted or desired. -->

* ([`r-explain`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-explain)) Pass with flying colors!
* ([`r-test`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-test)) Pass
* ~~([`r-code`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-code)) __Not evaluated__~~
* ([`r-doc`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-doc)) Pass
* ~~([`r-maint`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-maint)) __Not evaluated__~~
* ([`r-run`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-run)) :x: __Needs work__
* ([`r-user`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-user)) Pass
* ~~([`r-tech`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-tech)) __Not evaluated__~~

I ran this locally to see what I could break. I found the following issues:

* Can't select contact-type-limited relationships types when editing existing relationships
    1. View Relationships tab for an individual contact with a "Household Member of" relationship.
    1. Click "Edit" for the "Household Member of" relationship.
    1. **Expect** "Relationship Type" field to be set to "Household Member of".
    1. **Observe** "Relationship Type" field is empty and it's impossible to select "Household Member of".

* Too many relationship type options after saving a relationship type
    1. View Relationships tab for an individual.
    1. Click "Add Relationship".
    1. Click the "Edit Options" icon.
    1. On the "Employee of" type, click "Edit".
    1. Click "Save".
    1. Click to choose a relationship type.
    1. **Expect** types to be limited to valid relationships for individuals
    1. **Observe** option "Employer of" which is not valid for individuals.

* Recently enabled relationship type missing from options
    1. View Relationships tab for a contact with an existing relationship of a type which is disabled.
    1. Edit the relationship.
    1. Click the "Edit Options" icon.
    1. For the relationship type, click "More > Enable". Then click "Done".
    1. (Note: At this point it would be nice if "Relationship Type" could be populated with )
    1. Click to select a Relationship Type.
    1. **Expect** the relationship type which you just enabled to be a option.
    1. **Observe** that the relationship type is missing from the options.

mickadoo commented:
@seanmadsen thanks very much for the comprehensive QA

> Can't select contact-type-limited relationships types when editing existing relationships

Good catch. This seems to be related to the type casting I did. Unfortunately I found that unlike what the [doc block](https://github.com/civicrm/civicrm-core/blob/master/CRM/Contact/BAO/Relationship.php#L500) says `$contactId` is often a string value. I would prefer to have reliable types, but this function is used in more than 20 places just in core and I can't be sure of the type of `$contactId`, so I changed it to do loose comparison instead.

> Too many relationship type options after saving a relationship type

I couldn't reproduce this:

![peek 2018-03-27 14-56](https://user-images.githubusercontent.com/6374064/37971981-2f8166bc-31cf-11e8-9882-75bc4075fe87.gif)

In my GIF you might point out that I didn't edit the "A" side of the relationship. This is because (for me anyway) there seems to be a bug where editing the "A" side also replaces the value on the "B" side. I have created an [issue](https://issues.civicrm.org/jira/browse/CRM-21855) for this.

> Recently enabled relationship type missing from options

Another good spot. I had a look at this and found that whether the options are refreshed or not depends on if [the event returns data](https://github.com/civicrm/civicrm-core/blob/71bdf27c5e8d90d8de6e0a6b234f9da0128599a2/js/crm.ajax.js#L520). In [this commit](https://github.com/civicrm/civicrm-core/pull/11853/commits/8494007d67925582276552aba16ab26633e86fd5) I added some arbitrary data which will cause the options to be refreshed. As I mentioned in the commit message it is sort of a hack to get the desired functionality, but it could be useful data for reacting to the event.

One thing to note is that if you choose a disabled relationship type the option will be unselected. Even after enabling the relationship type in the popup it'll be unselected. This is because the rebuilding of the options just goes by the existing selected option, it doesn't do anything like fetching which option should be selected from the database. Basically if the option isn't selected already it doesn't know which one to select after.

seanmadsen commented:
Nice. Looking good, @mickadoo. I havn't re-tested, but I wanted to comment on the one you weren't able to reproduce... The steps I'm taking involve clicking on the "Edit" link, at the far right of the relationship type, which opens a separate form. Can you try that?

eileenmcnaughton commented:
@colemanw there is great review here by @seanmadsen but I think you also need to comment on the approach. Having 'contact_id' as a field for getoptions spec feels a bit odd having skimmed it - but I have not looked closely enough to have formed an opinion about it & the buildOptions structure is your brain child :-)

mickadoo commented:
@seanmadsen Ah, I didn't realize that. I tried to reproduce and actually I managed to first, but I noticed when I cleared the browser cache the problem went away. It might be an issue where the JS files are being cached, but the HTML content is showing the gear icon

![peek 2018-03-28 12-21](https://user-images.githubusercontent.com/6374064/38026155-ec54f232-3282-11e8-9611-807e1b580298.gif)

@eileenmcnaughton @colemanw A code review would be very helpful. I felt like I was going off the standard path here when I needed to make changes to how the spec functions are working, plus for other `getoptions` requests I don't usually see parameters to filter the available options. I just knew if I wanted to see the same options as in the original form after refreshing them I needed to pass the same criteria to the `getoptions` request.

seanmadsen commented:
Ah, caching! I see! Yes, I see what you mean. Ok...

I have reviewed this again after Michael's follow-up changes and can verify that he has addressed all my `r-run` concerns. 

Here's where we're at now then: 

* ([`r-explain`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-explain)) Pass
* ([`r-test`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-test)) Pass
* ~~([`r-code`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-code)) __Not evaluated__~~
* ([`r-doc`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-doc)) Pass
* ~~([`r-maint`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-maint)) __Not evaluated__~~
* ([`r-run`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-run)) Pass
* ([`r-user`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-user)) Pass
* ~~([`r-tech`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-tech)) __Not evaluated__~~

We still need someone to review those parts of the checklist that I skipped.

eileenmcnaughton commented:
Just as an aside I'm aware of another PR that is stalled really on needing core (& in particular Coleman to review). It's kind of similar in that there is a non-trivial amount of review effort. In that context I had a conversation about whether it would be possible to buy a 2 hour block of ct review time to get it prioritised over all the unfunded work (with no guarantee of merge in that time but at least definitely getting that time). It seems the blocker to such arrangements is often the 'sales process' - ie. the discussions around it take more time than is actually funded.

deb1990 commented on templates/CRM/Contact/Form/Relationship.tpl:
```diff
@@ -144,10 +144,113 @@
+          relationshipData = {},
+          contactTypes = {};
+
+        $('body').on('crmOptionsEdited', 'a.crm-option-edit-link', refreshRelationshipData);
```
Create a `init` function, it would help to organise the code, also place other initialisation code inside.
```
(function init () {
  $('body').on('crmOptionsEdited', 'a.crm-option-edit-link', refreshRelationshipData);
  refreshRelationshipData().done(...)
  $relationshipTypeSelect.change(function() {
})()
```

deb1990 commented on templates/CRM/Contact/Form/Relationship.tpl:
```diff
@@ -144,10 +144,113 @@
+         * Fetch contact types and reset relationship data
+         */
+        function refreshRelationshipData() {
+          var defer = $.Deferred();
```
No need to create `defer` object, the following will do the same thing
```javascript
relationshipData = {};

return getContactTypes()
  .then(function() {
    return getRelationshipData();
  });
});
```

deb1990 commented on templates/CRM/Contact/Form/Relationship.tpl:
```diff
@@ -144,10 +144,113 @@
+         * Fetches the relationship data using latest relationship types
+         */
+        function getRelationshipData() {
+          var subtype,
```
These variables are used only inside $.each, so we can define them inside only. Overall this is how it should look like
```javascript
function getRelationshipData() {
  var defer = $.Deferred();

  if (!$.isEmptyObject(relationshipData)) {
    defer.resolve(relationshipData);
  }

  CRM.api3("RelationshipType", "get")
    .done(function (data) {
      $.each(data.values, function (key, relType) {
        // Please add a comment mentioning what is ["a", "b"]
        $.each(["a", "b"], function (index, suffix) {
          var subtype = relType["contact_subtype_" + suffix];
          var type = subtype || relType["contact_type_" + suffix];
          var label = getContactTypeLabel(type) || "Contact";
          var placeholder = "- select " + label.toLowerCase() + " -";
          relType["placeholder_" + suffix] = placeholder;
        });

        relationshipData[relType["id"]] = relType;
      });

      defer.resolve(relationshipData);
    });

  return defer.promise();
}
```

mickadoo replied:
nice, I like the early return!

colemanw commented on templates/CRM/Contact/Form/Relationship.tpl:
```diff
@@ -144,10 +144,113 @@
+          }
+
+          return defer.promise();
+        }
```
I would recommend *not* loading contact types via ajax because the two back-to-back ajax calls will be unnecessarily slow. I suggest pre-loading contactType data into the dom (via `$form->assign()`) to avoid the lag.

mickadoo commented on CRM/Contact/Form/Relationship.php:
```diff
@@ -299,9 +299,6 @@ public function buildQuickForm() {
     $relationshipList = CRM_Contact_BAO_Relationship::getContactRelationshipType($this->_contactId, $this->_rtype, $this->_relationshipId);
 
-    // Metadata needed on clientside
-    $this->assign('relationshipData', self::getRelationshipTypeMetadata($relationshipList));
```
It would be nice to clean it up to avoid having code that does the same thing in JS and PHP, but it's still used in [one other place](https://github.com/civicrm/civicrm-core/blob/master/CRM/Case/Form/CaseView.php#L197). My JS skills are very limited, but if you think it's best that we remove this function and provide some common service to fetch this data then I can give that a shot.

colemanw replied:
Sorry, didn't realize it was used elsewhere. I don't think that's a priority.

mickadoo commented:
One test is failing, but seems to be the same on all PRs right now

colemanw commented:
Ok I think this is good to merge. I've completed the review started by Sean and it's now passing on all levels.

mickadoo commented:
Thanks for the in-depth review and QA @seanmadsen @colemanw and @deb1990 :slightly_smiling_face:

JKingsnorth commented:
Please see https://lab.civicrm.org/dev/core/issues/268

colemanw commented:
@mickadoo FYI the function `_civicrm_api3_relationship_getoptions_spec` has improper use of `ts()` - cannot use string concatenation within that function. I'm not even sure `ts()` is necessary in a field spec function since the strings are not user-facing.

eileenmcnaughton commented:
This gitlab has been logged https://lab.civicrm.org/dev/core/issues/895 - which suggests a regression that possibly sorta comes from this PR - but with other factors
